cpp/ql/src/experimental/Critical/MemoryMayNotBeFreedOnLoop.ql

```diff
@@ -0,0 +1,52 @@
+/**
+ * @name Memory May Not Be Freed On Loop
+ * @description A loop may break/continue before freeing memory that was allocated in the loop. A free/delete call should be added before the break/continue.
+ * @kind problem
+ * @id cpp/memory-may-not-be-freed-on-loop
+ * @problem.severity critical
+ * @tags efficiency
+ *       security
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.DataFlow
+import AllocAndFree
+
+predicate sameLoop(BlockStmt b, Stmt is) {
+  if is instanceof Loop then is.getAChild*() = b else sameLoop(b, is.getParent())
+}
+
+from StackVariable v, Expr def, JumpStmt b, BlockStmt bs, IfStmt is
+where
+  allocationDefinition(v, def) and
+  (b instanceof BreakStmt or b instanceof ContinueStmt) and
+  def.getParent() instanceof Initializer and
+  bs.getParent*() instanceof Loop and
+  exists(int i | i in [0 .. bs.getNumStmt() - 1] |
+    bs.getChild(i) = def.getEnclosingStmt() and
+    //Checking if the variable is not copied or used in a function call. If so, that might be normal if the variable is not freed in the loop.
+    not exists(int y | y in [0 .. bs.getNumStmt() - 1] |
+      bs.getChild(y) instanceof ExprStmt and
+      bs.getChild(y).(ExprStmt).getExpr() instanceof AssignExpr and
+      bs.getChild(y).(ExprStmt).getExpr().(Assignment).getRValue*() = v.getAnAccess()
+    ) and
+    exists(int y | y in [i .. bs.getNumStmt() - 1] |
+      (bs.getChild(y) = is or bs.getChild(y).(IfStmt).getAChild*() = is) and
+      sameLoop(bs, is) and
+      (
+        is.getAChild() = b or
+        is.getThen().getAChild() = b or
```

Comment:
```suggestion
        is.getThen().getAChild() = b
```

Reply:
Friendly ping @Yonah125 could you do these changes ?